File: src/aws/lambdas/incremental_distribution/src/test/java/uk/nhs/nhsx/core/routing/StandardHandlersTest.java
package uk.nhs.nhsx.core.routing;

import org.apache.http.entity.ContentType;
import org.junit.jupiter.api.Test;
import uk.nhs.nhsx.core.HttpResponses;
import uk.nhs.nhsx.core.TestEnvironments;
import uk.nhs.nhsx.core.exceptions.HttpStatusCode;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;
import static uk.nhs.nhsx.ProxyRequestBuilder.request;
import static uk.nhs.nhsx.matchers.ProxyResponseAssertions.hasStatus;

public class StandardHandlersTest {

    @Test
    public void oaiHeaderIsRequiredToMatchWhenEnvironmentVariableIsSet_Set() throws Exception {

        Map<String, String> env = new HashMap<>() {{
            put("custom_oai", "bob");
        }};

        Routing.Handler handler = StandardHandlers.requiringCustomAccessIdentity(TestEnvironments.TEST.apply(env), r -> HttpResponses.ok());

        assertThat(handler.handle(request().build()), hasStatus(HttpStatusCode.FORBIDDEN_403));
        assertThat(handler.handle(request().withHeader("x-custom-oai", "bob").build()), hasStatus(HttpStatusCode.OK_200));
        assertThat(handler.handle(request().withHeader("x-custom-oai", "jim").build()), hasStatus(HttpStatusCode.FORBIDDEN_403));
        assertThat(handler.handle(request().withHeader("x-custom-oai", "").build()), hasStatus(HttpStatusCode.FORBIDDEN_403));
        assertThat(handler.handle(request().withHeader("x-custom-oai-something", "bob").build()), hasStatus(HttpStatusCode.FORBIDDEN_403));
    }

    @Test
    public void oaiHeaderIsRequiredToMatchWhenEnvironmentVariableIsSet_NotSet() throws Exception {

        Routing.Handler handler = StandardHandlers.requiringCustomAccessIdentity(TestEnvironments.EMPTY, r -> HttpResponses.ok());

        assertThat(handler.handle(request().build()), hasStatus(HttpStatusCode.OK_200));
        assertThat(handler.handle(request().withHeader("x-custom-oai", "bob").build()), hasStatus(HttpStatusCode.OK_200));
    }

    @Test
    public void requiringAuthorizationHeader_supplied() throws Exception {
        Routing.Handler handler = StandardHandlers.requiringAuthorizationHeader(r -> HttpResponses.ok());
        assertThat(handler.handle(request().withHeader("authorization", "something").build()), hasStatus(HttpStatusCode.OK_200));
    }

    @Test
    public void requiringAuthorizationHeader_notSupplied() throws Exception {
        Routing.Handler handler = StandardHandlers.requiringAuthorizationHeader(r -> HttpResponses.ok());
        assertThat(handler.handle(request().build()), hasStatus(HttpStatusCode.FORBIDDEN_403));
    }

    @Test
    public void authorisedBy_notSupplied() throws Exception {
        Routing.Handler handler = StandardHandlers.authorisedBy(authorizationHeader -> true, r -> HttpResponses.ok());
        assertThat(handler.handle(request().build()), hasStatus(HttpStatusCode.FORBIDDEN_403));
    }

    @Test
    public void authorisedBy_authorised() throws Exception {
        Routing.Handler handler = StandardHandlers.authorisedBy(authorizationHeader -> true, r -> HttpResponses.ok());
        assertThat(handler.handle(request().withHeader("authorization", "something").build()), hasStatus(HttpStatusCode.OK_200));
    }

    @Test
    public void authorisedBy_notAuthorised() throws Exception {
        Routing.Handler handler = StandardHandlers.authorisedBy(authorizationHeader -> false, r -> HttpResponses.ok());
        assertThat(handler.handle(request().withHeader("authorization", "something").build()), hasStatus(HttpStatusCode.FORBIDDEN_403));
    }

    @Test
    public void expectingContentType_suppliedCorrect() throws Exception {
        Routing.Handler handler = StandardHandlers.expectingContentType(ContentType.APPLICATION_JSON, r -> HttpResponses.ok());
        assertThat(handler.handle(request().withHeader("Content-Type", "application/json").build()), hasStatus(HttpStatusCode.OK_200));
        assertThat(handler.handle(request().withHeader("Content-Type", "application/json;charset=utf-8").build()), hasStatus(HttpStatusCode.OK_200));
    }

    @Test
    public void expectingContentType_suppliedIncorrect() throws Exception {
        Routing.Handler handler = StandardHandlers.expectingContentType(ContentType.APPLICATION_JSON, r -> HttpResponses.ok());
        assertThat(handler.handle(request().withHeader("Content-Type", "text/html").build()), hasStatus(HttpStatusCode.UNPROCESSABLE_ENTITY_422));
        assertThat(handler.handle(request().withHeader("Content-Type", "application/vnd.ms-excel").build()), hasStatus(HttpStatusCode.UNPROCESSABLE_ENTITY_422));
    }

    @Test
    public void expectingContentType_notSupplied() throws Exception {
        Routing.Handler handler = StandardHandlers.expectingContentType(ContentType.APPLICATION_JSON, r -> HttpResponses.ok());
        assertThat(handler.handle(request().build()), hasStatus(HttpStatusCode.BAD_REQUEST_400));
    }

    @Test
    public void filteringWhileMaintenanceModeEnabled_suppliedTrue() throws Exception {
        Map<String, String> env = Map.of("MAINTENANCE_MODE", "TRUE");
        Routing.Handler handler = StandardHandlers.filteringWhileMaintenanceModeEnabled(TestEnvironments.TEST.apply(env), r -> HttpResponses.ok());
        assertThat(handler.handle(request().build()), hasStatus(HttpStatusCode.SERVICE_UNAVAILABLE_503));
    }

    @Test
    public void filteringWhileMaintenanceModeEnabled_suppliedFalse() throws Exception {
        Map<String, String> env = Map.of("MAINTENANCE_MODE", "FALSE");
        Routing.Handler handler = StandardHandlers.filteringWhileMaintenanceModeEnabled(TestEnvironments.TEST.apply(env), r -> HttpResponses.ok());
        assertThat(handler.handle(request().build()), hasStatus(HttpStatusCode.OK_200));
    }

    @Test
    public void filteringWhileMaintenanceModeEnabled_suppliedEmpty() throws Exception {
        Map<String, String> env = Map.of("MAINTENANCE_MODE", "");
        Routing.Handler handler = StandardHandlers.filteringWhileMaintenanceModeEnabled(TestEnvironments.TEST.apply(env), r -> HttpResponses.ok());
        assertThat(handler.handle(request().build()), hasStatus(HttpStatusCode.OK_200));
    }

    @Test
    public void handlesEmptyAuthHeader() {
        String authorizationHeader = "";
        assertThat(StandardHandlers.apiKeyNameFrom(authorizationHeader), is(Optional.empty()));
    }

    @Test
    public void handlesEmptyApiKey() {
        String apiKey = "";
        String authorizationHeader = "Bearer " + apiKey;
        assertThat(StandardHandlers.apiKeyNameFrom(authorizationHeader), is(Optional.empty()));
    }

    @Test
    public void handlesApiKeyWithNonBase64Encoding() {
        String apiKey = "name:value";
        String authorizationHeader = "Bearer " + apiKey;
        assertThat(StandardHandlers.apiKeyNameFrom(authorizationHeader), is(Optional.empty()));
    }



}

File: src/aws/lambdas/incremental_distribution/src/test/java/uk/nhs/nhsx/core/routing/StandardHandlersTest.kt
package uk.nhs.nhsx.core.routing

import com.amazonaws.HttpMethod.GET
import com.amazonaws.services.lambda.runtime.Context
import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyRequestEvent
import org.apache.http.entity.ContentType
import org.hamcrest.CoreMatchers.equalTo
import org.hamcrest.CoreMatchers.not
import org.hamcrest.MatcherAssert.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.fail
import uk.nhs.nhsx.core.HttpResponses
import uk.nhs.nhsx.core.TestEnvironments
import uk.nhs.nhsx.core.auth.ResponseSigner
import uk.nhs.nhsx.core.events.IncomingHttpRequest
import uk.nhs.nhsx.core.events.RecordingEvents
import uk.nhs.nhsx.core.events.RequestRejected
import uk.nhs.nhsx.core.exceptions.HttpStatusCode.BAD_REQUEST_400
import uk.nhs.nhsx.core.exceptions.HttpStatusCode.FORBIDDEN_403
import uk.nhs.nhsx.core.exceptions.HttpStatusCode.OK_200
import uk.nhs.nhsx.core.exceptions.HttpStatusCode.SERVICE_UNAVAILABLE_503
import uk.nhs.nhsx.core.exceptions.HttpStatusCode.UNPROCESSABLE_ENTITY_422
import uk.nhs.nhsx.core.headers.MobileAppVersion
import uk.nhs.nhsx.core.headers.MobileOS.Android
import uk.nhs.nhsx.core.headers.MobileOSVersion
import uk.nhs.nhsx.core.headers.UserAgent
import uk.nhs.nhsx.core.routing.Routing.RouterMatch.Matched
import uk.nhs.nhsx.core.routing.StandardHandlers.authorisedBy
import uk.nhs.nhsx.core.routing.StandardHandlers.expectingContentType
import uk.nhs.nhsx.core.routing.StandardHandlers.filteringWhileMaintenanceModeEnabled
import uk.nhs.nhsx.core.routing.StandardHandlers.loggingIncomingRequests
import uk.nhs.nhsx.core.routing.StandardHandlers.requiringAuthorizationHeader
import uk.nhs.nhsx.core.routing.StandardHandlers.requiringCustomAccessIdentity
import uk.nhs.nhsx.core.routing.StandardHandlers.signedBy
import uk.nhs.nhsx.core.routing.StandardHandlers.withoutSignedResponses
import uk.nhs.nhsx.testhelper.ContextBuilder.TestContext
import uk.nhs.nhsx.testhelper.ProxyRequestBuilder.Companion.request
import uk.nhs.nhsx.testhelper.matchers.ProxyResponseAssertions.hasHeader
import uk.nhs.nhsx.testhelper.matchers.ProxyResponseAssertions.hasStatus
import java.time.Instant

class StandardHandlersTest {

    private val testContext = TestContext()
    private  val events = RecordingEvents()

    @Test
    fun `sign successful responses`() {
        val signer = ResponseSigner { _, response -> response.headers["signature"] = "signature" }
        val handler = signedBy(signer) { _, _ -> HttpResponses.ok() }
        val response = handler(request().build(), testContext)
        assertThat(response, hasHeader("signature", equalTo("signature")))
    }

    @Test
    fun `doesn't sign 403 responses`() {
        val signer = ResponseSigner { _, response -> response.headers["signature"] = "signature" }
        val handler = signedBy(signer) { _, _ -> HttpResponses.withStatusCode(FORBIDDEN_403) }
        val response = handler(request().build(), testContext)
        assertThat(response, not(hasHeader("signature")))
    }

    @Test
    fun `oai header is required to match when environment variable is set`() {
        val env = mapOf("custom_oai" to "bob")
        val handler = requiringCustomAccessIdentity(
            RecordingEvents(),
            TestEnvironments.TEST.apply(env)
        ) { _, _ -> HttpResponses.ok() }
        assertThat(handler(request().build(), testContext), hasStatus(FORBIDDEN_403))
        assertThat(handler(request().withHeader("x-custom-oai", "bob").build(), testContext), hasStatus(OK_200))
        assertThat(handler(request().withHeader("x-custom-oai", "jim").build(), testContext), hasStatus(FORBIDDEN_403))
        assertThat(handler(request().withHeader("x-custom-oai", "").build(), testContext), hasStatus(FORBIDDEN_403))
        assertThat(
            handler(request().withHeader("x-custom-oai-something", "bob").build(), testContext), hasStatus(
                FORBIDDEN_403
            )
        )
    }

    @Test
    fun `requiring authorizationHeader supplied`() {
        val handler = requiringAuthorizationHeader { _, _ -> HttpResponses.ok() }
        assertThat(handler(request().withHeader("authorization", "something").build(), testContext), hasStatus(OK_200))
    }

    @Test
    fun `requiring AuthorizationHeader not supplied`() {
        val handler = requiringAuthorizationHeader { _, _ -> HttpResponses.ok() }
        assertThat(handler(request().build(), testContext), hasStatus(FORBIDDEN_403))
    }

    @Test
    fun `authorisedBy notSupplied`() {
        val handler = authorisedBy({ true }, okRoutingHandler())
        assertThat(handler(request().build(), testContext), hasStatus(FORBIDDEN_403))
    }

    private fun okRoutingHandler() = object : Routing.RoutingHandler {
        override fun match(request: APIGatewayProxyRequestEvent) = Matched(this)
        override fun invoke(request: APIGatewayProxyRequestEvent, context: Context) = HttpResponses.ok()
    }

    @Test
    fun `authorisedBy authorised`() {
        val handler = authorisedBy({ true }, okRoutingHandler())
        assertThat(handler(request().withHeader("authorization", "something").build(), testContext), hasStatus(OK_200))
    }

    @Test
    fun `authorisedBy notAuthorised`() {
        val handler = authorisedBy({ false }, okRoutingHandler())
        assertThat(
            handler(request().withHeader("authorization", "something").build(), testContext),
            hasStatus(FORBIDDEN_403)
        )
    }

    @Test
    fun `authorisedBy match and handle`() {
        val request = request().build()
        val handler = authorisedBy({ true }, okRoutingHandler())

        when (val match = handler.match(request)) {
            is Matched -> assertThat(match.handler(request, testContext), hasStatus(FORBIDDEN_403))
            else -> fail { "match is not of type Matched" }
        }
    }

    @Test
    fun `expectingContentType suppliedCorrect`() {
        val handler = expectingContentType(ContentType.APPLICATION_JSON) { _, _ -> HttpResponses.ok() }
        assertThat(
            handler(request().withHeader("Content-Type", "application/json").build(), testContext),
            hasStatus(OK_200)
        )
        assertThat(
            handler(request().withHeader("Content-Type", "application/json;charset=utf-8").build(), testContext),
            hasStatus(
                OK_200
            )
        )
    }

    @Test
    fun `expectingContentType suppliedIncorrect`() {
        val handler = expectingContentType(ContentType.APPLICATION_JSON) { _, _ -> HttpResponses.ok() }
        assertThat(
            handler(request().withHeader("Content-Type", "text/html").build(), testContext), hasStatus(
                UNPROCESSABLE_ENTITY_422
            )
        )
        assertThat(
            handler(request().withHeader("Content-Type", "application/vnd.ms-excel").build(), testContext), hasStatus(
                UNPROCESSABLE_ENTITY_422
            )
        )
    }

    @Test
    fun `expectingContentType notSupplied`() {
        val handler = expectingContentType(ContentType.APPLICATION_JSON) { _, _ -> HttpResponses.ok() }
        assertThat(handler(request().build(), testContext), hasStatus(BAD_REQUEST_400))
    }

    @Test
    fun `filtering while maintenance mode enabled supplied True`() {
        val env = mapOf("MAINTENANCE_MODE" to "TRUE")
        val handler =
            filteringWhileMaintenanceModeEnabled(events, TestEnvironments.TEST.apply(env)) { _, _ -> HttpResponses.ok() }
        assertThat(handler(request().build(), testContext), hasStatus(SERVICE_UNAVAILABLE_503))
        events.containsExactly(RequestRejected::class)
    }

    @Test
    fun `filtering while maintenance mode enabled supplied False`() {
        val env = mapOf("MAINTENANCE_MODE" to "FALSE")
        val handler =
            filteringWhileMaintenanceModeEnabled(events, TestEnvironments.TEST.apply(env)) { _, _ -> HttpResponses.ok() }
        assertThat(handler(request().build(), testContext), hasStatus(OK_200))
        events.containsExactly()
    }

    @Test
    fun `filtering while maintenance mode enabled supplied Empty`() {
        val env = mapOf("MAINTENANCE_MODE" to "")
        val handler =
            filteringWhileMaintenanceModeEnabled(events, TestEnvironments.TEST.apply(env)) { _, _ -> HttpResponses.ok() }
        assertThat(handler(request().build(), testContext), hasStatus(OK_200))
        events.containsExactly()
    }

    @Test
    fun `without signed response`() {
        val events = RecordingEvents()
        val environment = TestEnvironments.TEST.apply(
            mapOf(
                "custom_oai" to "OAI",
                "MAINTENANCE_MODE" to "FALSE"
            )
        )

        val handler = withoutSignedResponses(events, environment) { _, _ -> HttpResponses.ok() }

        val request = request()
            .withMethod(GET)
            .withPath("/hello")
            .withCustomOai("OAI")
            .withHeader("User-Agent", "Android")
            .build()

        handler(request, testContext)

        events.containsExactly(IncomingHttpRequest::class)
    }

    @Test
    fun `logs incoming requests`() {
        val events = RecordingEvents()

        val handler = loggingIncomingRequests(events, { _, _ -> HttpResponses.ok() }, { Instant.EPOCH })

        val request = request()
            .withMethod(GET)
            .withPath("/hello")
            .withCustomOai("OAI")
            .withHeader("User-Agent", "p=Android,o=29,v=4.3.0,b=138")
            .build()

        handler(request, testContext)

        val incoming = events.first() as IncomingHttpRequest
        assertThat(incoming.uri, equalTo("/hello"))
        assertThat(incoming.method, equalTo("GET"))
        assertThat(incoming.apiKey, equalTo("none"))
        assertThat(
            incoming.userAgent,
            equalTo(UserAgent(MobileAppVersion.Version(4, 3, 0), Android, MobileOSVersion.of("29")))
        )
        assertThat(incoming.latency, equalTo(0))
        assertThat(incoming.requestId, equalTo("none"))
        assertThat(incoming.status, equalTo(200))
        assertThat(incoming.message, equalTo("Received http request: method=GET,path=/hello,requestId=none,apiKeyName=none,userAgent=p=Android,o=29,v=4.3.0,b=138,status=200,latency=PT0S"))
    }
}
